core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -929,6 +955,28 @@ class DAGScheduler(
       // HadoopRDD whose underlying HDFS files have been deleted.
       finalStage = createResultStage(finalRDD, func, partitions, jobId, callSite)
     } catch {
+      case e: Exception if e.getMessage ==
+          DAGScheduler.ERROR_MESSAGE_BARRIER_REQUIRE_MORE_SLOTS_THAN_CURRENT_TOTAL_NUMBER =>
+        logWarning("The job requires to run a barrier stage that requires more slots than the " +
+          "total number of slots in the cluster currently.")
+        jobIdToNumTasksCheckFailures.putIfAbsent(jobId, 0)
+        val numCheckFailures = jobIdToNumTasksCheckFailures.get(jobId) + 1
+        if (numCheckFailures < DAGScheduler.DEFAULT_MAX_CONSECUTIVE_NUM_TASKS_CHECK_FAILURES) {
```

Comment:
Should make `DEFAULT_MAX_CONSECUTIVE_NUM_TASKS_CHECK_FAILURES` configurable so users can specify unlimited retry if needed. Instead, we might want to fix the timeout since it is only relevant to cost.